doc/developer/design/20230721_persist_fast_path.md

```diff
@@ -0,0 +1,81 @@
+
+Why no filters?
+
+- Index-based fast-path queries are often used for single-key lookups, like `select * from indexed_collection where id = 'XXX'`. Persist cannot currently support efficient queries with this shape.
+- In general, a query with a filter may need to do unbounded work to generate a finite result set size. (In the limit, we might need to scan the entire Persist shard to generate a result set of only a few rows!) If we disallow filters, we can effectively bound the amount of work we assign to `clusterd` by bounding the result set size.
+
+Today, every select query generates a peek request into an arrangement. (Even non-fast-path queries work by creating a new dataflow that writes out its data into an arrangement, then peeking into that arrangement.) This design introduces a second type of peek: one that peeks directly into the backing Persist shard. Concrentely, this means that `environmentd` will send `ComputeCommand::Peek` commands that reference a persisted collection’s `Id`, not the id of an index or arrangement.
+
+### Peeking
+
+When `clusterd` receives a peek command, it stores and tracks a `PendingPeek` struct with the arrangement and request metadata in the compute state, polls until that arrangement has data for the requested timestamp, then read that data synchronously back out of the arrangement before returning a `ComputeResponse::Peek` with the results.
+
+Since Persist’s API is asynchronous, one straightforward approach would be to fork off a new task for every incoming Peek, and hold on to the task handle. We can periodically poll the task for completion, and cancel it if the peek itself is cancelled by the user.
+
+When executing an ordinary peek, each worker reads a subset of the results out of their local arrangement… so much of the work is spread equally across all workers/processes of a replica. However, with Persist-backed peeks, we’d really only be doing the reading in *one* task running on a single node. We’ll need to take care to ensure these peek tasks are fairly distributed to avoid unduly skewing the workload.
+
+### Persist
+
+For background on the relevant bits of Persist, see [the background section of the consolidation-on-read design doc](https://github.com/bkirwi/materialize/blob/consolidation-on-read/doc/developer/design/20230317_consolidation_on_read.md#background). That section mentions:
+
+> A single worker can efficiently and fully consolidate a set of runs using a streaming merge: iterat[ing] through each of the runs concurrently, pulling from whichever run has the smallest next key-value, and consolidating updates with equal key-value-time as you go.
+>
+
+We can’t actually get away with this in the distributed context, since the Persist source is spread across many workers. However, for a little peek, running on a single node should be fine… and this approach is more or less what we intend to implement.
```

Comment:
> [...] for a little peek, running on a single node [...]

Just to check my understanding here: When you say _node_ above, are you referring to a single replica process, potentially in a multi-process replica?

Reply:
Specifically I meant running on a single node per replica: as opposed to running on every node or worker in a replica.

I do not have a strong opinion on whether such a query should be targeted at just one replica in a cluster or at every replica. My understanding is that existing peeks go to every replica (is that right?) and I don't have any reason to deviate from the existing behaviour one way or another in that respect... but I suspect that either approach could be made to work.

Reply:
Sounds good; by default peeks go to every replica, but it is possible to perform replica-targeted peeks AFAIR (ref https://github.com/MaterializeInc/materialize/issues/12520). So we can explore what makes more sense depending on the characteristics of Persist-backed, fast-path peeks.